File: internal/scan/cache.go
package scan

import (
	"encoding/json"
	"os"
	"path/filepath"

	"houston/internal/accounts"
	"houston/internal/model"
)

// Cache remembers the parsed Mission of every transcript keyed by path and
// validated by (size, mtime), so TUI launches and rescans only re-read files
// that actually changed — parsing streams, but a big store still costs tens of
// MB of I/O per pass. It's a pure accelerator: any load/save problem simply
// means a full re-parse. Note the cached Cwd/HasSubagents reflect the disk
// state at parse time; resume re-checks the dir and fails gracefully if it
// moved.
type Cache struct {
	path    string
	entries map[string]cacheEntry
	used    map[string]bool // paths seen this run; Save prunes the rest
	dirty   bool
}

type cacheEntry struct {
	Size    int64         `json:"size"`
	MTimeNS int64         `json:"mtimeNs"`
	Mission model.Mission `json:"mission"`
}

// LoadCache reads the scan cache from Houston's data dir (an empty cache if
// missing or unreadable).
func LoadCache() *Cache {
	c := &Cache{
		path:    filepath.Join(accounts.StoreDir(), "scan-cache.json"),
		entries: map[string]cacheEntry{},
		used:    map[string]bool{},
	}
	if b, err := os.ReadFile(c.path); err == nil {
		var on struct {
			Entries map[string]cacheEntry `json:"entries"`
		}
		if json.Unmarshal(b, &on) == nil && on.Entries != nil {
			c.entries = on.Entries
		}
	}
	return c
}

// lookup returns the cached mission if the file is unchanged. Nil-safe so the
// uncached Scan path can share the walker.
func (c *Cache) lookup(path string, fi os.FileInfo) (model.Mission, bool) {
	if c == nil || fi == nil {
		return model.Mission{}, false
	}
	e, ok := c.entries[path]
	if !ok || e.Size != fi.Size() || e.MTimeNS != fi.ModTime().UnixNano() {
		return model.Mission{}, false
	}
	c.used[path] = true
	return e.Mission, true
}

func (c *Cache) put(path string, fi os.FileInfo, m model.Mission) {
	if c == nil || fi == nil {
		return
	}
	c.entries[path] = cacheEntry{Size: fi.Size(), MTimeNS: fi.ModTime().UnixNano(), Mission: m}
	c.used[path] = true
	c.dirty = true
}

// Save persists the entries touched this run — pruning transcripts that no
// longer exist — atomically and best-effort.
func (c *Cache) Save() {
	if c == nil || c.path == "" {
		return
	}
	if !c.dirty && len(c.used) == len(c.entries) {
		return // clean pass over the exact same set: nothing to write
	}
	kept := map[string]cacheEntry{}
	for p := range c.used {
		if e, ok := c.entries[p]; ok {
			kept[p] = e
		}
	}
	b, err := json.Marshal(struct {
		Entries map[string]cacheEntry `json:"entries"`
	}{kept})
	if err != nil {
		return
	}
	if os.MkdirAll(filepath.Dir(c.path), 0o700) != nil {
		return
	}
	tmp, err := os.CreateTemp(filepath.Dir(c.path), "scan-cache-*.tmp")
	if err != nil {
		return
	}
	name := tmp.Name()
	if _, err := tmp.Write(b); err != nil {
		tmp.Close()
		os.Remove(name)
		return
	}
	if tmp.Close() != nil {
		os.Remove(name)
		return
	}
	if os.Rename(name, c.path) != nil {
		os.Remove(name)
	}
}

File: crates/houston-core/src/scan/cache.rs
//! The scan cache: remembers the parsed Mission of every transcript keyed by
//! path and validated by (size, mtime), so TUI launches and rescans only
//! re-read files that actually changed. A pure accelerator: any load/save
//! problem simply means a full re-parse. Same on-disk SHAPE as v1 but its own
//! FILE (`<store>/scan-cache-v2.json`): while v1 and v2 coexist on the same
//! machine, neither may clobber the other's warm cache.

use crate::model::Mission;
use crate::paths::store_dir;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Serialize, Deserialize, Clone)]
struct Entry {
    size: u64,
    #[serde(rename = "mtimeNs")]
    mtime_ns: i64,
    mission: Mission,
}

#[derive(Serialize, Deserialize, Default)]
struct OnDisk {
    entries: HashMap<String, Entry>,
}

pub struct Cache {
    path: PathBuf,
    entries: HashMap<String, Entry>,
    /// Paths seen this run; save() prunes the rest.
    used: HashSet<String>,
    dirty: bool,
}

impl Cache {
    /// Read the scan cache from Houston's data dir (empty if missing/unreadable).
    pub fn load() -> Self {
        Self::load_from(store_dir().join("scan-cache-v2.json"))
    }

    pub fn load_from(path: PathBuf) -> Self {
        let entries = fs::read(&path)
            .ok()
            .and_then(|b| serde_json::from_slice::<OnDisk>(&b).ok())
            .map(|d| d.entries)
            .unwrap_or_default();
        Cache { path, entries, used: HashSet::new(), dirty: false }
    }

    /// The cached mission if the file is unchanged.
    pub fn lookup(&mut self, path: &Path, fi: &fs::Metadata) -> Option<Mission> {
        let key = path.to_string_lossy().into_owned();
        let e = self.entries.get(&key)?;
        if e.size != fi.len() || e.mtime_ns != super::mtime_ns(fi) {
            return None;
        }
        self.used.insert(key);
        Some(e.mission.clone())
    }

    pub fn put(&mut self, path: &Path, fi: &fs::Metadata, m: &Mission) {
        let key = path.to_string_lossy().into_owned();
        self.entries.insert(
            key.clone(),
            Entry { size: fi.len(), mtime_ns: super::mtime_ns(fi), mission: m.clone() },
        );
        self.used.insert(key);
        self.dirty = true;
    }

    /// Persist, pruning entries whose files were not seen this run. Best-effort.
    pub fn save(&mut self) {
        let pruned = self.entries.len() != self.used.len();
        if !self.dirty && !pruned {
            return;
        }
        self.entries.retain(|k, _| self.used.contains(k));
        let disk = OnDisk { entries: std::mem::take(&mut self.entries) };
        if let Some(dir) = self.path.parent() {
            let _ = fs::create_dir_all(dir);
        }
        if let Ok(b) = serde_json::to_vec(&disk) {
            // Uniquely-named temp: a fixed ".tmp" lets two processes writing at
            // once (two TUIs, or a TUI and `houston scan`) interleave into the
            // same file and rename half a document into place. The cache is only
            // an accelerator, but a corrupt one costs a full re-parse of every
            // transcript on the next launch.
            let tmp = self.path.with_extension(format!("json.{}.tmp", std::process::id()));
            if fs::write(&tmp, b).is_ok() {
                if fs::rename(&tmp, &self.path).is_err() {
                    let _ = fs::remove_file(&tmp);
                }
            } else {
                let _ = fs::remove_file(&tmp);
            }
        }
        self.entries = disk.entries;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    #[test]
    fn cache_roundtrip_hit_and_invalidation() {
        let tmp = tempfile::tempdir().unwrap();
        let file = tmp.path().join("s.jsonl");
        let mut f = fs::File::create(&file).unwrap();
        writeln!(f, r#"{{"type":"user","message":{{"content":"hi"}}}}"#).unwrap();
        drop(f);

        let cache_path = tmp.path().join("cache.json");
        let mut c = Cache::load_from(cache_path.clone());
        let fi = fs::metadata(&file).unwrap();
        assert!(c.lookup(&file, &fi).is_none(), "cold cache must miss");

        let m = crate::scan::parse_file(&file).unwrap();
        c.put(&file, &fi, &m);
        assert!(c.lookup(&file, &fi).is_some(), "hot cache must hit");
        c.save();

        // Reload: still hits for the unchanged file.
        let mut c2 = Cache::load_from(cache_path.clone());
        assert!(c2.lookup(&file, &fi).is_some(), "persisted cache must hit");

        // Grow the file: size changes → miss.
        let mut f = fs::OpenOptions::new().append(true).open(&file).unwrap();
        writeln!(f, r#"{{"type":"user","message":{{"content":"more"}}}}"#).unwrap();
        drop(f);
        let fi2 = fs::metadata(&file).unwrap();
        let mut c3 = Cache::load_from(cache_path);
        assert!(c3.lookup(&file, &fi2).is_none(), "changed file must miss");
    }

    #[test]
    fn save_prunes_unseen_entries() {
        let tmp = tempfile::tempdir().unwrap();
        let file = tmp.path().join("s.jsonl");
        fs::write(&file, "{}\n").unwrap();
        let cache_path = tmp.path().join("cache.json");

        let mut c = Cache::load_from(cache_path.clone());
        let fi = fs::metadata(&file).unwrap();
        c.put(&file, &fi, &Mission::default());
        c.save();

        // Next run never touches the file → save() drops it.
        let mut c2 = Cache::load_from(cache_path.clone());
        c2.dirty = true;
        c2.save();
        let c3 = Cache::load_from(cache_path);
        assert!(c3.entries.is_empty(), "unseen entries must be pruned");
    }
}
